sri/templatetags/sri.py

```diff
@@ -10,42 +11,72 @@
 from sri.algorithm import DEFAULT_ALGORITHM, Algorithm
 from sri.integrity import calculate_integrity_of_static
 
+
 USE_SRI = getattr(settings, "USE_SRI", not settings.DEBUG)
 
 register = template.Library()
 
 
-def sri_js(attrs: dict, path: str, algorithm: Algorithm):
-    attrs.update({"type": "text/javascript", "src": static(path)})
-    return mark_safe(f"<script{flatatt(attrs)}></script>")
+def format_attrs(*simple_attrs, **complex_attrs) -> str:
+    """
+    Flatten and format list and dict params.
+
+    The returned value will be a string with a leading " " that is a
+    combination of "name=value" and "name" params:
+
+        ' src="/" defer'
+        ' href="/" rel="stylesheet"'
+
+    """
+    complex = flatatt(complex_attrs)
+    if simple_attrs:
+        simple = " ".join(simple_attrs)
+        return f"{complex} {simple}".rstrip()
+    return complex.rstrip()
+
+
+def sri_js(path: str, *simple_attrs: str, **complex_attrs: str) -> str:
+    complex_attrs.setdefault("src", static(path))
+    return script_tag(path, *simple_attrs, **complex_attrs)
+
 
+def sri_css(path: str, *simple_attrs: str, **complex_attrs: str) -> str:
+    complex_attrs.setdefault("rel", "stylesheet")
+    complex_attrs.setdefault("type", "text/css")
+    return link_tag(path, *simple_attrs, **complex_attrs)
 
-def sri_css(attrs: dict, path: str, algorithm: Algorithm):
-    attrs.update({"rel": "stylesheet", "type": "text/css", "href": static(path)})
-    return mark_safe(f"<link{flatatt(attrs)}/>")
+
+def sri_font(path: str, *simple_attrs: str, **complex_attrs: str) -> str:
+    complex_attrs.setdefault("rel", "stylesheet")
+    complex_attrs.setdefault("type", "text/css")
+    return link_tag(path, *simple_attrs, **complex_attrs)
+
+
+def script_tag(path: str, *simple_attrs: str, **complex_attrs: str) -> str:
+    return mark_safe(f"<script{format_attrs(*simple_attrs, **complex_attrs)}></script>")
+
+
+def link_tag(path: str, *simple_attrs: str, **complex_attrs: str) -> str:
+    complex_attrs.setdefault("href", static(path))
+    return mark_safe(f"<link{format_attrs(*simple_attrs, **complex_attrs)}>")
 
 
 EXTENSIONS = {"js": sri_js, "css": sri_css}
 
 
 @register.simple_tag
-def sri_static(path: str, algorithm: Optional[str] = None):
-    algorithm_type = Algorithm(algorithm or DEFAULT_ALGORITHM)
+def sri_static(
+    path: str, *simple_attrs: str, algorithm: str | None = None, **complex_attrs: str
```

Comment:
Suggestion: It would be nice to require the algorithm be first before the `simple_attrs` so the order is more deterministic. Alternatively, forcing `algorithm` to be a kwarg could also be nice and explicit.

Reply:
This is an interesting one - when I was looking at it I couldn't think of a scenario where you would use different algorithms within the same project (i.e. SHA-256 for some files, SHA-384 for others). Is this a real use case?

There are three possible forms:

```python
# least favourite as you have to be explicit on *every* use
def sri_static(path: str, algorithm: str, *simple_attrs, **complex_attrs) -> str:
    """Algorithm as required argument."""


# second least favourite - it's neat, but it does mix function kwargs with tag attrs.
def sri_static(path: str, *simple_attrs, **complex_attrs) -> str:
    """Algorithm read out from kwargs."""
    algorithm = complex_attrs.pop("algorithm", DEFAULT_ALGORITHM.value)


# personal favourite - it's explicit, it doesn't mix with tag attrs, it's optional
def sri_static(path: str, *simple_attrs, algorithm: str = DEFAULT_ALOGITHM.value, **complex_attrs) -> str:
    """Algorithm as named kwarg."""

```

Reply:
My thinking was more:

```python
def sri_static(path: str, algorithm: str = DEFAULT_ALOGITHM.value, *simple_attrs, **complex_attrs) -> str:
```

Calling becomes the same, but it prevents `{% sri_static "index.js" "defer" algorithm="sha512" %}` which looks messy

Reply:
If you specify `algorithm` before `*args` it will always interpret the first unnamed argument as the algorithm, so you have to include it if you have any other unnamed arguments:

```python
>>> def sri_static(path, algo = "sha256", *args, **kwargs):
...     print(f"path:  {path}\nalgo:  {algo}\nargs:  {args}\nkwargs: {kwargs}")
...
```
```python
>>> sri_static("index.js")
path:  index.js
algo:  sha256
args:  ()
kwargs: {}
```
```python
>>> sri_static("index.js", "defer")
path:  index.js
algo:  defer
args:  ()
kwargs: {}
```
```python
>>> sri_static("index.js", "defer", algo="sha384")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sri_static() got multiple values for argument 'algo'
>>>
```

Reply:
I guess it comes down to use cases.

Selfishly, we have hundreds of static includes and none of them would require the algorithm to be explicit as we only use the same `DEFAULT_ALGORITHM` for everything, but we do have a whole smorgasbord of attrs (empty and not empty).

Reply:
Yeah I agree that's not ideal. Perhaps we should enforce that `algorithm` be a keyword argument?

Reply:
Unfortunately you can't have positional args after kwargs - which means you'd have to make _everything_ a kwarg, which then means that you need to store a list of "empty attr" names so that you can treat them differently.

Reply:
Perhaps something like:

```python
def sri_static(path: str, *simple_attrs, *, algorithm: str = DEFAULT_ALOGITHM.value, **complex_attrs) -> str:
```

Reply:
@RealOrangeOne this is what it looks like after the last update?

Reply:
It does, almost. The main change is the extra `*, ` forcing `algorithm` to be specified as a keyword argument.

Reply:
Ah yes - good spot - apologies - updating now.

Reply:
Turns out you can't do that:

> Keyword-only argument separator not allowed after "*" parameter